### PR TITLE
Implement Streamlabels (Except trains)

### PR DESCRIPTION
**Pull request by andycreeth**
This PR implements streamlabels in the app.

Here are some details:
- Initial fetch of data is done via HTTP request
- Subsequent updates are handled via websocket events
- Custom properties on properties managers have been ditched in favor of custom UI components to handle properties.  This is because OBS property types were too restrictive to implement what we wanted for streamlabels.

### CODE REVIEW

File: app/services/streamlabels/definitions.ts [hunk 1]
```diff
@@ -0,0 +1,861 @@
+import { TPlatform } from 'services/platforms';
```
Comment (andycreeth) [inline app/services/streamlabels/definitions.ts]: Most of this file is copied from the existing streamlabels app.  I munged the data a bit, removed mixer definitions for now, and defined some interfaces based on the structure of the data.

File: app/components/windows/AddSource.vue.ts [hunk 2]
```diff
@@ -65,11 +67,16 @@ export default class AddSource extends Vue {
...
       this.error = 'That name is already taken';
     } else {
-      const sceneItem = this.scenesService.activeScene.createAndAddSource(
+      const source = this.sourcesService.createSource(
         this.name,
-        this.sourceType
+        this.sourceType,
+        {},
+        {
+          propertiesManager: this.propertiesManager ? this.propertiesManager : void 0
```
Comment (holiber) [inline app/components/windows/AddSource.vue.ts]: Do we really need explicitly set which property manager we have to use? What will happen if I try to use `Default` property manager for `widget` or `streamlabels` ?

File: app/services/streamlabels/index.ts [hunk 3]
```diff
@@ -0,0 +1,212 @@
...
+        if (e.type === 'streamlabels') {
+          this.setStreamlabelsData(e.message.data);
+        }
+      });
+    });
+  }
+
+
+  log(message: string, ...args: any[]) {
+    console.log(`Streamlabels: ${message}`, ...args);
```
Comment (holiber) [inline app/services/streamlabels/index.ts]: do we need logs always enabled?

File: app/services/streamlabels/index.ts [hunk 4]
```diff
@@ -0,0 +1,212 @@
...
+      .then(handleErrors)
+      .then(() => true);
+  }
+
+
+  /**
+   * Attempt to load initial data via HTTP instead of waiting
+   * for a socket event
+   */
+  fetchInitialData() {
```
Comment (holiber) [inline app/services/streamlabels/index.ts]: can you set returning type for this function?

File: app/services/streamlabels/index.ts [hunk 5]
```diff
@@ -0,0 +1,212 @@
...
+    return path.join(this.streamlabelsDirectory, `${filename}.txt`);
+  }
+
+
+  /**
+   * Updates the data in a single streamlabels file
+   * @param filename the filename
+   * @param statname the name of the stat
+   */
+  updateSubscription(subscription: IStreamlabelsSubscription) {
```
Comment (holiber) [inline app/services/streamlabels/index.ts]: I don't see params above as arguments for this metod

File: app/services/streamlabels/index.ts [hunk 6]
```diff
@@ -0,0 +1,212 @@
...
+    if (!this.data) return;
+
+    console.log('Writing file', subscription.filename, subscription.statname);
```
Comment (holiber) [inline app/services/streamlabels/index.ts]: should be `this.log()`

File: app/services/streamlabels/index.ts [hunk 7]
```diff
@@ -0,0 +1,212 @@
...
+
+    electron.ipcRenderer.send('streamlabels-writeFile', {
+      path: this.getStreamlabelsPath(subscription.filename),
+      data: this.data[subscription.statname]
+    });
```
Comment (holiber) [inline app/services/streamlabels/index.ts]: Why we need main process here?

File: app/services/streamlabels/index.ts (hunk 3 above)
Comment (andycreeth) [reply app/services/streamlabels/index.ts]: Yeah maybe I should add an ENV var

File: app/components/windows/AddSource.vue.ts (hunk 2 above)
Comment (andycreeth) [reply app/components/windows/AddSource.vue.ts]: Yes.  The properties manager determines whether it is streamlabels or not.  For example:
- A "normal" text source: `text_gdiplus` source with `default` manager
- A "Stream Label": `text_gdiplus` source with `streamlabel` manager
- A "normal" browser source: `browser_source` with `default` manager
- A "Widget": `browser_source` with `widget` manager

File: app/services/streamlabels/index.ts (hunk 5 above)
Comment (andycreeth) [reply app/services/streamlabels/index.ts]: Oops, I changed the method and forgot to edit the comment

File: app/services/streamlabels/index.ts (hunk 6 above)
Comment (andycreeth) [reply app/services/streamlabels/index.ts]: Actually meant to remove this one entirely.

File: app/services/streamlabels/index.ts (hunk 7 above)
Comment (andycreeth) [reply app/services/streamlabels/index.ts]: In theory, we shouldn't need the main process at all.  Unfortunately there are a lot of issues doing asynchronous writes from the renderer process.  Synchronous writes work just fine (we do those for config saving and a few other things).  However, asynchronous writes will fail a good portion of the time.  Given the high volume of writes we need to do for streamlabels, I wasn't comfortable doing them synchronously.  Asynchronous writes work fine from the main process, so I am using an asynchronous IPC to do them from there.

As best I can tell, the issue seems to be a bug in electron.  I lost many hours to investigation and it turns out that Karl hit the exact same issue when writing the original streamlabels app (also electron).  When we have time we are going to try to reproduce in a test case and file a bug with electron.

Comment (holiber) [reply app/services/streamlabels/index.ts]: Ouch

**Pull request merged**